Implement new MessageLogger configuration

Dr15Jones opened the pull request:
#### PR description:

This adds support for using the new MessageLogger configuration syntax. The old syntax is also supported. The decision on which format to apply is based on the presence or absence of key parameters for each implementation.

The discussion about the new format can be found here: #31910

The new code makes use of the ParameterSet validation system to enforce correctness.

The standard MessageLogger_cfi.py configuration has not been updated to use the new configuration at this point since it would cause too many non framework configurations to break. That migration will happen at a later stage.

#### PR validation:

All framework unit tests have been updated to use the new syntax and they all pass.

Dr15Jones committed:
Specify use of statistics directly in destinations

This removes the need for the statistics vstring as the specification of directing the statistics to a given destination is done via the 'enableStatistics' bool parameter.

Dr15Jones committed:
Follow CMS member data naming conventions

Dr15Jones committed:
Do not hold PSet as member data

Dr15Jones committed:
Implement new MessageLogger configuration API

-Both the old and new configuration APIs are supported. Having `categories` and `destinations` parameters being available trigger the old configuration handling.
-As much code as reasonably possible is shared between the two implementations.
-The ParameterSetValidation code is used for the new API but the fillDescriptions usage can not be implemented until all uses of the old configuration are removed.

Dr15Jones committed:
Fixed formatting

Dr15Jones committed:
Fix new configuration problems

Properly read the 'files' PSet
Set the correct default for lineLength
Make it so the 'default' PSet can properly set the values for destinations which do not explicitly set a parameter.

Dr15Jones committed:
Member data names now conform to CMS standard

Dr15Jones committed:
Correct problems when using new MessageLogger configuration

-Properly handle setting up statistics
-Set defaults to match original behavior
-Throw an exception if multiple destinations want to write to the same log file

Dr15Jones committed:
Updated MessageLogger tests to use new configuration syntax

Only the tests which explicitly created the Service were changed. Those that use the standard cfi.py file must wait until we are ready to change that file.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32138/19807

- This PR adds an extra 220KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32138/19807/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32138/19807/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

Dr15Jones committed:
code-checks update

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32138/19808

- This PR adds an extra 220KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Framework
FWCore/MessageService
FWCore/ParameterSet


@makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
Please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 315a8f9c5d508856c6d6a6821a004ba0bc86ffa3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bfd43c/10750/summary.html
CMSSW: CMSSW_11_2_X_2020-11-13-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bfd43c/10750/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2529267
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 148 log files, 22 edm output root files, 35 DQM output files

makortel commented on FWCore/ParameterSet/interface/ParameterWildcardWithSpecifics.h:
```diff
@@ -0,0 +1,42 @@
+  class ParameterSet;
+  class DocFormatHelper;
+
+  class ParameterWildcardWithSpecifics : public ParameterWildcardBase {
```
Could you document the usage of this class in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Wildcards_More_Details
after this PR has been merged?

Dr15Jones replied:
OK

Dr15Jones commented on FWCore/MessageService/src/ThreadSafeLogMessageLoggerScribe.cc:
```diff
@@ -687,22 +875,128 @@ namespace edm {
+    destination_base.addUntracked<edm::ParameterSetDescription>("FWKINFO",category);
+    destination_base.addUntracked<edm::ParameterSetDescription>("WARNING",category);
+    destination_base.addUntracked<edm::ParameterSetDescription>("ERROR",category);
+    */
```
Nope. I have actually removed it in my own copy already.

Dr15Jones commented on FWCore/MessageService/test/filter-timestamps.sed:
```diff
@@ -1 +1 @@
-s/ [0-9]?[0-9]-(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-2[0-9][0-9][0-9] [0-9]?[0-9]:[0-9][0-9]:[0-9][0-9](\.[0-9][0-9][0-9])? [A-Z]?[A-Z][A-Z][A-Z]( |\t|$)/ {Timestamp} /
+s/ [0-9]?[0-9]-(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-2[0-9][0-9][0-9] [0-9]?[0-9]:[0-9][0-9]:[0-9][0-9](\.[0-9][0-9][0-9])? [A-Z]?[A-Z][A-Z][A-Z]( |\t|-|$)/ {Timestamp} /
```
One of the unit tests was failing because of an extra minus sign that happened in the time printout at its end. I can't remember exactly which one. I looked into and determined the code was doing the right thing and the filtering was just missing a corner case.

Dr15Jones commented on FWCore/MessageService/src/ThreadSafeLogMessageLoggerScribe.cc:
```diff
@@ -687,22 +875,128 @@ namespace edm {
+
+        edm::ParameterWildcard<edm::ParameterSetDescription> catnode("*", edm::RequireZeroOrMore, false, category);
+        catnode.setComment("Specialize either a category or any of 'DEBUG', 'INFO', 'FWKINFO', 'WARNING' or 'ERROR'");
+        //destination_base.addWildcardUntracked<edm::ParameterSetDescription>(category);
```
Nope

Dr15Jones committed:
Removed old code based on singleton pattern

The ELadministrator has not been a singleton for a long time. The old code was preventing the use of std::make_shared with the class.

Dr15Jones committed:
Code improvements based on a code-review

Dr15Jones commented on FWCore/MessageService/src/ThreadSafeLogMessageLoggerScribe.cc:
```diff
@@ -35,16 +38,13 @@ namespace edm {
-          ,
+        : m_admin_p(new ELadministrator()),
+          m_early_dest(m_admin_p->attach(std::make_shared<ELoutput>(std::cerr, false))),
+          m_file_ps(),
```
done

Dr15Jones commented on FWCore/MessageService/src/ThreadSafeLogMessageLoggerScribe.cc:
```diff
@@ -189,174 +190,380 @@ namespace edm {
+        for (auto const& name : files.getParameterNamesForType<edm::ParameterSet>(false)) {
+          categories.merge(findCategoriesInDestination(files.getUntrackedParameter<edm::ParameterSet>(name)));
+        }
+        categories.merge(std::set<std::string>(psets.begin(), psets.end()));
```
done

Dr15Jones commented on FWCore/MessageService/src/ThreadSafeLogMessageLoggerScribe.cc:
```diff
@@ -189,174 +190,380 @@ namespace edm {
+      // Attach a default extension of .log if there is no extension on a file
+      // change log 18 - this had been done in concert with attaching destination
+
+      std::string actual_filename = filename;  // change log 4
```
done

Dr15Jones commented:
All code changes have been made.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32138/19903

- This PR adds an extra 164KB to repository

cmsbuild commented:
Pull request #32138 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: cbec163780d32f1951bd6f22b62665501ecad5ea
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bfd43c/10848/summary.html
CMSSW: CMSSW_11_2_X_2020-11-18-1600
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bfd43c/10848/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2529267
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 148 log files, 34 edm output root files, 35 DQM output files

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
